### PR TITLE
Adjust nullability of a target type for an argument of a lifted binary operator.

### LINKED ISSUE #39361: Incorrect 'CS8629 Nullable value type may be null' on implicitly-cast operand of lifted operator
**Version Used**: New in VS 16.4p2, not present in 16.4p1 or 16.3.5

No warning should be shown because the operator is lifted and therefore expects nulls:

```cs
#nullable enable

struct Foo
{
    public static implicit operator DateTime?(Foo? value) => default;

    bool M(Foo? x, DateTime y)
    {
        // CS8629 Nullable value type may be null.
        //     ↓
        return x == y;
    }
}
```

The other operators also trigger this bug.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #39361.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @jcouv, @dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@cston, @jcouv, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -3160,11 +3160,18 @@ private void AfterLeftChildHasBeenVisited(BoundExpression leftOperand, Conversio
                     ParameterSymbol parameter,
                     TypeWithState operandType)
                 {
+                    TypeWithAnnotations targetTypeWithNullability = parameter.TypeWithAnnotations;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3163]: >targetTypeWithNullability [](start = 40, length = 25)

Nit: Not a fan of the naming. `Nullability` in this file generally means either annotation state or flow state. This is really `possiblyLiftedTargetType`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -121902,6 +121902,82 @@ class Program
...
+    {
+        return x5 == y5;
+    }
+
+    bool M6(S? x6, DateTime? y6)
+    {
+        return x6 == y6;
+    }
+
+    bool M7(DateTime? x7, S? y7)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:121949]: Consider adding a case for `DateTime x8, S y8`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3163]: The name matches parameter name of the VisitConversion method that is called below.

---
In reply to: [405826407](https://github.com/dotnet/roslyn/pull/43165#discussion_r405826407) [](ancestors = 405826407)

**Pull request merged**